### ISSUE TITLE
MultiLayerNetwork.setParameters Broken

### ISSUE THREAD

**Issue by aurelienwaite** [deeplearning4j/deeplearning4j#533]
In MultiLayerNetwork.setParameters the line

idx += range - 1;

should be 

idx += range;

It's causing the parameters to be offset. For example 

[ -0.01, 0.04, -0.07, 0.08, 0.02, 0.00, -0.36, -0.03, 0.11, -0.04, 0.14, 0.12, 0.07, 0.16, -0.09, 0.12, -0.02, -0.01, -0.06, 0.04, 0.00, 0.00, 0.03, 0.02, 0.17, -0.09, 0.01, -0.15, 0.01, 0.04, 0.09, -0.28, 0.11, -0.00, -0.02, -0.22, 0.04, 0.04, 0.19, 0.03, 0.01, 0.17, -0.02, 0.04, -0.03, 0.06, -0.01, 0.05, 0.05, 0.19, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, -0.01, 0.04, -0.07, 0.08, 0.02, 0.00, -0.36, -0.03, 0.11, -0.04, 0.00, 0.00]

Is transformed into 

[ -0.01, 0.04, -0.07, 0.08, 0.02, 0.00, -0.36, -0.03, 0.11, -0.04, 0.14, 0.12, 0.07, 0.16, -0.09, 0.12, -0.02, -0.01, -0.06, 0.04, 0.00, 0.00, 0.03, 0.02, 0.17, -0.09, 0.01, -0.15, 0.01, 0.04, 0.09, -0.28, 0.11, -0.00, -0.02, -0.22, 0.04, 0.04, 0.19, 0.03, 0.01, 0.17, -0.02, 0.04, -0.03, 0.06, -0.01, 0.05, 0.05, 0.19, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, -0.01, 0.04, -0.07, 0.08, 0.02, 0.00, -0.36, -0.03, 0.11, -0.04, 0.00]

Below is a reproduction that can be run in the scala console

import org.deeplearning4j.nn.conf.distribution.NormalDistribution
import org.deeplearning4j.nn.weights.WeightInit
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork
import org.deeplearning4j.nn.conf.NeuralNetConfiguration
import org.deeplearning4j.nn.conf.layers.DenseLayer
import org.deeplearning4j.nn.conf.layers.OutputLayer
import org.nd4j.linalg.lossfunctions.LossFunctions
import org.deeplearning4j.nn.layers.BaseLayer
import org.nd4j.linalg.factory.Nd4j
import org.deeplearning4j.nn.params.PretrainParamInitializer

val conf = new NeuralNetConfiguration.Builder().seed(3l).activationFunction("relu") .weightInit(WeightInit.DISTRIBUTION).dist(new NormalDistribution(0, 1e-1)).iterations(1).corruptionLevel(0.0).dropOut(0.0).list(2).layer(0, new DenseLayer.Builder().nIn(10).nOut(5).build()).layer(1, new OutputLayer.Builder(LossFunctions.LossFunction.XENT).activation("softmax").nIn(5).nOut(2).build() ).backward(true).pretrain(false).build()
  
val nn = new MultiLayerNetwork(conf)
nn.init
  
  //Why do I have to do the following? Shouldn't this be unnecessary after the init? Is this another bug?
val layer = nn.getLayer(0).asInstanceOf[BaseLayer]
layer.setInput(Nd4j.ones(10))
layer.setParam(PretrainParamInitializer.VISIBLE_BIAS_KEY, Nd4j.zeros(10))
layer.conf().addVariable(PretrainParamInitializer.VISIBLE_BIAS_KEY)
  
  //Bug below
val params = nn.params
nn.setParameters(params)
val params2 = nn.params
println(params.equals(params2))
println(params)
println(params2)

**Comment by agibsonccc** [comment]
This is likely related to nd4j's indexing issues. I'm going to wait till after the fix to look at this.

**Comment by agibsonccc** [comment]
This should be fixed now. Please re open if you run in to this again (just update nd4j)

**Comment by aurelienwaite** [comment]
I've done an update and tried the reproduction. It now throws the following exception when trying to set the parameters

scala> nn.setParameters(params)
java.lang.ArrayIndexOutOfBoundsException: 1
  at org.nd4j.linalg.indexing.Indices.offsets(Indices.java:148)
  at org.nd4j.linalg.indexing.ShapeOffsetResolution.exec(ShapeOffsetResolution.java:35)
  at org.nd4j.linalg.api.ndarray.BaseNDArray.get(BaseNDArray.java:3754)
  at org.deeplearning4j.nn.multilayer.MultiLayerNetwork.setParameters(MultiLayerNetwork.java:1561)
  ... 43 elided

Could you reopen this issue? I don't have access rights?

**Comment by agibsonccc** [comment]
This is fixed now.
